File: api/v1/endpoints/tasks.py
from typing import List, Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from db.database import get_db
from services.task_service import TaskService
from schemas.task import (
    Task, TaskCreate, TaskUpdate, TaskWithWeather,
    TaskComment, TaskCommentCreate, TaskList
)
from models.task import TaskStatus, TaskCategory

router = APIRouter()

@router.post("/tasks", response_model=Task)
async def create_task(
    task_data: TaskCreate,
    db: Session = Depends(get_db)
):
    """
    Crée une nouvelle tâche avec ses ressources et dépendances associées.
    """
    task_service = TaskService(db)
    return await task_service.create_task(task_data)

@router.get("/tasks/{task_id}", response_model=Task)
def get_task(
    task_id: int,
    db: Session = Depends(get_db)
):
    """
    Récupère une tâche par son ID.
    """
    task_service = TaskService(db)
    return task_service.get_task(task_id)

@router.get("/tasks/{task_id}/weather", response_model=TaskWithWeather)
async def get_task_with_weather(
    task_id: int,
    db: Session = Depends(get_db)
):
    """
    Récupère une tâche avec analyse des conditions météo actuelles.
    """
    task_service = TaskService(db)
    return await task_service.get_task_with_weather(task_id)

@router.put("/tasks/{task_id}", response_model=Task)
def update_task(
    task_id: int,
    task_data: TaskUpdate,
    db: Session = Depends(get_db)
):
    """
    Met à jour une tâche existante.
    """
    task_service = TaskService(db)
    return task_service.update_task(task_id, task_data)

@router.delete("/tasks/{task_id}")
def delete_task(
    task_id: int,
    db: Session = Depends(get_db)
):
    """
    Supprime une tâche et libère ses ressources.
    """
    task_service = TaskService(db)
    task_service.delete_task(task_id)
    return {"message": "Tâche supprimée avec succès"}

@router.get("/projects/{project_id}/tasks", response_model=TaskList)
def get_project_tasks(
    project_id: int,
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    status: Optional[TaskStatus] = None,
    category: Optional[TaskCategory] = None,
    db: Session = Depends(get_db)
):
    """
    Récupère les tâches d'un projet avec pagination et filtres.
    """
    task_service = TaskService(db)
    return task_service.get_tasks_by_project(
        project_id=project_id,
        page=page,
        size=size,
        status=status,
        category=category
    )

@router.post("/tasks/{task_id}/comments", response_model=TaskComment)
def add_task_comment(
    task_id: int,
    comment_data: TaskCommentCreate,
    db: Session = Depends(get_db)
):
    """
    Ajoute un commentaire à une tâche.
    """
    task_service = TaskService(db)
    return task_service.add_task_comment(comment_data)

@router.put("/tasks/{task_id}/resources/{resource_id}")
def update_resource_usage(
    task_id: int,
    resource_id: int,
    quantity_used: float = Query(..., gt=0),
    db: Session = Depends(get_db)
):
    """
    Met à jour l'utilisation d'une ressource pour une tâche.
    """
    task_service = TaskService(db)
    task_service.update_task_resource(task_id, resource_id, quantity_used)
    return {"message": "Utilisation de la ressource mise à jour"}

@router.get("/tasks/{task_id}/dependencies", response_model=List[Task])
def get_task_dependencies(
    task_id: int,
    db: Session = Depends(get_db)
):
    """
    Récupère toutes les tâches qui dépendent de la tâche spécifiée.
    """
    task_service = TaskService(db)
    return task_service.get_dependent_tasks(task_id)

@router.get("/tasks/weather-dependent", response_model=List[TaskWithWeather])
async def get_weather_dependent_tasks(
    db: Session = Depends(get_db)
):
    """
    Récupère toutes les tâches dépendantes de la météo avec leur statut actuel.
    """
    task_service = TaskService(db)
    return await task_service.get_weather_dependent_tasks()

File: frontend/src/services/tasks.ts
import { Task, TaskFormData, TaskWithWeather, TaskList } from '../types/task';

const BASE_URL = '/api/v1';

export const getTasks = async (
  projectId: number,
  page: number = 1,
  status?: string,
  category?: string
): Promise<TaskList> => {
  const params = new URLSearchParams({
    page: page.toString()
  });

  if (status) params.append('status', status);
  if (category) params.append('category', category);

  const response = await fetch(
    `${BASE_URL}/projects/${projectId}/tasks?${params.toString()}`
  );

  if (!response.ok) {
    throw new Error('Erreur lors de la récupération des tâches');
  }

  return response.json();
};

export const getTask = async (taskId: number): Promise<Task> => {
  const response = await fetch(`${BASE_URL}/tasks/${taskId}`);

  if (!response.ok) {
    throw new Error('Erreur lors de la récupération de la tâche');
  }

  return response.json();
};

export const getTaskWeather = async (taskId: number): Promise<TaskWithWeather> => {
  const response = await fetch(`${BASE_URL}/tasks/${taskId}/weather`);

  if (!response.ok) {
    throw new Error('Erreur lors de la récupération des données météo');
  }

  return response.json();
};

export const createTask = async (taskData: TaskFormData): Promise<Task> => {
  const response = await fetch(`${BASE_URL}/tasks`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(taskData)
  });

  if (!response.ok) {
    throw new Error('Erreur lors de la création de la tâche');
  }

  return response.json();
};

export const updateTask = async (taskId: number, taskData: TaskFormData): Promise<Task> => {
  const response = await fetch(`${BASE_URL}/tasks/${taskId}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(taskData)
  });

  if (!response.ok) {
    throw new Error('Erreur lors de la mise à jour de la tâche');
  }

  return response.json();
};

export const deleteTask = async (taskId: number): Promise<void> => {
  const response = await fetch(`${BASE_URL}/tasks/${taskId}`, {
    method: 'DELETE'
  });

  if (!response.ok) {
    throw new Error('Erreur lors de la suppression de la tâche');
  }
};

export const getWeatherDependentTasks = async (): Promise<TaskWithWeather[]> => {
  const response = await fetch(`${BASE_URL}/tasks/weather-dependent`);

  if (!response.ok) {
    throw new Error('Erreur lors de la récupération des tâches météo-dépendantes');
  }

  return response.json();
};

export const updateTaskResource = async (
  taskId: number,
  resourceId: number,
  quantityUsed: number
): Promise<void> => {
  const response = await fetch(
    `${BASE_URL}/tasks/${taskId}/resources/${resourceId}?quantity_used=${quantityUsed}`,
    {
      method: 'PUT'
    }
  );

  if (!response.ok) {
    throw new Error("Erreur lors de la mise à jour de l'utilisation des ressources");
  }
};
